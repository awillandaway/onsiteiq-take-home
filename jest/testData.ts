import type { Candidate } from '../src/types/Candidate';

export const mockCandidate: Candidate = {
  gender: 'male',
  name: {
    title: 'Mr',
    first: 'John',
    last: 'Smith',
  },
  location: {
    street: {
      number: 1,
      name: 'Test Street',
    },
    city: 'New York',
    state: 'New York',
    country: 'United States',
    postcode: 11111,
    coordinates: {
      latitude: 'test-latitude',
      longitude: 'test-longitude',
    },
    timezone: {
      offset: '-5:00',
      description: 'US East',
    },
  },
  email: 'john.smith@test.com',
  dob: {
    date: '1955-11-04T21:17:18.030Z',
    age: 67,
  },
  registered: {
    date: '2017-10-15T15:36:03.501Z',
    age: 5,
  },
  phone: '(123) 456-7890',
  cell: '(111) 111-1111',
  id: {
    name: 'drivers license',
    value: '11111111111',
  },
  picture: {
    large: 'https://randomuser.me/api/portraits/men/76.jpg',
    medium: 'https://randomuser.me/api/portraits/med/men/76.jpg',
    thumbnail: 'https://randomuser.me/api/portraits/thumb/men/76.jpg',
  },
  nat: 'US',
  status: 'not_yet_reviewed',
  savedId: 'test-not-reviewed-id',
};

export const mockApprovedCandidate: Candidate = {
  ...mockCandidate,
  status: 'approved',
  savedId: 'test-approved-id',
  notes: 'this candidate is great',
};

export const mockRejectedCandidate: Candidate = { ...mockCandidate, status: 'rejected', savedId: 'test-rejected-id' };
