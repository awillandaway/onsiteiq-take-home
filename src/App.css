body {
  margin: 0;
  background-color: rgb(37, 37, 37);
  color: #fff;
}

* {
  font-family: 'Roboto', sans-serif;
}

h1, h2, h3, h4, h5, h6, p {
  margin: 0;
}