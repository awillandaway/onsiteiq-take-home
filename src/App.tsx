import { useModal } from 'hooks/useModal';
import { useState } from 'react';
import { getRandomCandidate } from 'services/CandidateService';
import { CandidateInfoCard } from 'components/CandidateInfoCard/CandidateInfoCard';
import type { Candidate } from 'types/Candidate';
import { MainContent, Navbar, NotificationNumber, NotificationsSection } from './App.styles';
import './App.css';

const App = () => {
  const { isOpen, toggleModal } = useModal();
  const [currentCandidateInfo, setCurrentCandidateInfo] = useState<Candidate | null>();

  const onClickGetCandidate = () => {
    getRandomCandidate().then((candidate) => {
      setCurrentCandidateInfo(candidate);
    });
  };

  return (
    <div>
      <Navbar>
        <h1>OnsiteIQ HR Application</h1>
        <NotificationsSection>
          You have <NotificationNumber>{Math.floor(Math.random() * 8) + 2}</NotificationNumber> candidates to review.
        </NotificationsSection>
      </Navbar>

      <MainContent>
        {/** add icons to buttons */}
        <button type="button" onClick={onClickGetCandidate}>
          Start New Candidate Review
        </button>
        <button type="button">View Completed Reviews</button> {/** this may make sense to open a separate page */}
        <pre>{JSON.stringify(currentCandidateInfo)}</pre>
        {!currentCandidateInfo && <div>Please click the Start New Candidate Review button to start a review.</div>}
        {currentCandidateInfo && <CandidateInfoCard candidate={currentCandidateInfo} />}
      </MainContent>
    </div>
  );
};

export default App;
