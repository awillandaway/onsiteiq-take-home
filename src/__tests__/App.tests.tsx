import { fireEvent, render, waitForElementToBeRemoved } from '@testing-library/react';
import renderer from 'react-test-renderer';
import { mockApprovedCandidate, mockCandidate, mockRejectedCandidate } from '../../jest/testData';
import App from '../App';

const mockCandidates = JSON.parse(localStorage.getItem('candidates') as string);

jest.mock('src/services/CandidateService', () => ({
  getRandomCandidate: jest.fn(
    () =>
      new Promise((resolve) => {
        resolve(mockCandidate);
      }),
  ),
  getSavedCandidates: jest.fn(
    () =>
      new Promise((resolve) => {
        resolve(mockCandidates);
      }),
  ),
  saveCandidate: jest.fn(),
  updateCandidate: jest.fn(),
}));

describe('App', () => {
  it('Renders correctly', () => {
    const app = renderer.create(<App />).toJSON();
    expect(app).toMatchSnapshot();
  });

  it('Renders the Start New Review section on initial render', async () => {
    const { findByText, findByRole } = render(<App />);

    expect(await findByText('Please click the Start New Candidate Review button to start a review.')).toBeVisible();
    expect(await findByRole('button', { name: 'Start New Candidate Review' })).toBeVisible();
  });

  it('Loads saved (reviewed) candidates on render', async () => {
    const { findAllByTestId } = render(<App />);

    expect(JSON.parse(localStorage.getItem('candidates') as string)).toStrictEqual([
      mockApprovedCandidate,
      mockRejectedCandidate,
    ]);

    const candidateInfoCards = await findAllByTestId('candidate-info-card');

    expect(candidateInfoCards).toHaveLength(2);
  });

  it('Clicking the Start New Candidate Review button replaces the Start New Review section with a candidate to review', async () => {
    const { findAllByTestId, queryByText, findByRole } = render(<App />);

    const startNewReviewButton = await findByRole('button', { name: 'Start New Candidate Review' });

    fireEvent.click(startNewReviewButton);

    await waitForElementToBeRemoved(() =>
      queryByText('Please click the Start New Candidate Review button to start a review.'),
    );

    const candidateInfoCards = await findAllByTestId('candidate-info-card');

    expect(candidateInfoCards).toHaveLength(3); // 1 card for the new review, 2 cards for the saved reviews
  });
});
