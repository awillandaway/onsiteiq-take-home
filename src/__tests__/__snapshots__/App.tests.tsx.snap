// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App Renders correctly 1`] = `
<div>
  <nav
    className="sc-csuSiG hvMVPV"
  >
    <h1>
      OnsiteIQ HR Application
    </h1>
    <div
      className="sc-eDWCr eTiBAU"
    >
      You have 
      <span
        className="sc-bqWxrE dWqmdV"
      >
        9
      </span>
       candidates to review.
    </div>
  </nav>
  <div
    className="sc-ksBlkl bgcEsR"
  >
    <div
      className="sc-fEXmlR juCLJa"
    >
      <h2>
        New Review
      </h2>
      <div
        className="sc-fnGiBr bogsAz"
      >
        Please click the Start New Candidate Review button to start a review.
        <button
          className="sc-bcXHqe jWfLtu"
          onClick={[Function]}
          type="button"
        >
          Start New Candidate Review
        </button>
      </div>
    </div>
    <div
      className="sc-bjfHbI bUWTEh"
    >
      <h2>
        Saved Reviews
      </h2>
      <div
        className="sc-hBxehG kRbnFu"
      />
    </div>
  </div>
</div>
`;
